app/scripts/controllers/permissions/loggerMiddleware.js

```diff
@@ -81,51 +81,58 @@ module.exports = function createLoggerMiddleware ({
     return Object.keys(request.params[0])
   }
 
-  function logHistory (requestedMethods, origin, result, time) {
-
-    let accounts
-    const entries = result
-      ? result.map(perm => {
-        if (perm.parentCapability === 'eth_accounts') {
-          accounts = getAccountsFromPermission(perm)
-        }
-        return perm.parentCapability
-      })
-        .reduce((acc, m) => {
-          if (requestedMethods.includes(m)) {
-            acc[m] = {
-              lastApproved: time,
-            }
-            if (m === 'eth_accounts') {
-              acc[m].accounts = accounts
-            }
+  function logHistory (requestedMethods, origin, result, time, isEthRequestAccounts) {
+    let accounts, entries
+    if (isEthRequestAccounts) {
+      accounts = result
+      const accountToTimeMap = accounts.reduce((acc, account) => ({ ...acc, [account]: time }), {})
+      entries = { 'eth_accounts': { accounts: accountToTimeMap, lastApproved: time } }
+    } else {
+      entries = result
+        ? result.map(perm => {
+          if (perm.parentCapability === 'eth_accounts') {
+            accounts = getAccountsFromPermission(perm)
           }
-          return acc
-        }, {})
-      : {}
+          return perm.parentCapability
+        })
+          .reduce((acc, m) => {
+            if (requestedMethods.includes(m)) {
+              if (m === 'eth_accounts') {
+                const accountToTimeMap = accounts.reduce((acc, account) => ({ ...acc, [account]: time }), {})
+                acc[m] = { lastApproved: time, accounts: accountToTimeMap }
+              } else {
+                acc[m] = { lastApproved: time }
+              }
+            }
+            return acc
+          }, {})
+        : {}
+    }
 
     if (Object.keys(entries).length > 0) {
-      commitHistory(origin, entries, accounts)
+      commitHistory(origin, entries)
     }
   }
 
-  function commitHistory (origin, entries, accounts) {
-
-    const history = store.getState()[historyStoreKey]
+  function commitHistory (origin, entries) {
+    const history = store.getState()[historyStoreKey] || {}
+    const newOriginHistory = {
+      ...history[origin],
+      ...entries,
+    }
 
-    if (Array.isArray(accounts)) {
-      if (history[origin] && history[origin]['eth_accounts']) {
-        history[origin]['eth_accounts']['accounts']
-          .filter(acc => !accounts.includes(acc))
-          .forEach(acc => accounts.unshift(acc))
+    if (history[origin] && history[origin]['eth_accounts'] && entries['eth_accounts']) {
+      newOriginHistory['eth_accounts'] = {
+        lastApproved: entries['eth_accounts'].lastApproved,
+        accounts: {
+          ...history[origin]['eth_accounts'].accounts,
```

Comment:
Ah I see - so the change in behavior here is that accounts that _used_ to be approved but are no longer are now preserved in this history? They'll just have an older timestamp. Whereas earlier they were overwritten.

Reply:
Correct. This is to allow us to know the last connect time per account.